### PR TITLE
DEPR/REGR: Fix pandas.util.testing deprecation

### LINKED ISSUE #30735: BUG/DEPR: the deprecation of util.testing fails for direct imports
```
In [1]: from pandas.util.testing import assert_frame_equal  
---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
<ipython-input-1-79d99d902fdd> in <module>
----> 1 from pandas.util.testing import assert_frame_equal

ImportError: cannot import name 'assert_frame_equal' from 'pandas.util.testing' (/home/joris/scipy/pandas/pandas/util/testing/__init__.py)
```

It works when accessing from top-level import:

```
In [3]: pd.util.testing.assert_frame_equal
/home/joris/miniconda3/envs/dev/bin/ipython:1: FutureWarning: pandas._testing.assert_frame_equal is deprecated. Please use pandas.testing.assert_frame_equal instead.
  #!/home/joris/miniconda3/envs/dev/bin/python
Out[3]: <function pandas._testing.assert_frame_equal(left, right, check_dtype=True, check_index_type='equiv', check_column_type='equiv', check_frame_type=True, check_less_precise=False, check_names=True, by_blocks=False, check_exact=False, check_datetimelike_compat=False, check_categorical=True, check_like=False, obj='DataFrame')>
```

### PR DESCRIPTION
**Pull request by TomAugspurger**
Closes https://github.com/pandas-dev/pandas/issues/30735

This avoids using _DeprecatedModule, which doesn't work for
direct imports from a module. Sorry for the importlib magic, but
I think this is the correct way to do things.

cc @jorisvandenbossche.

### CODE REVIEW

File: pandas/util/__init__.py [hunk 1]
```diff
@@ -1,4 +1,60 @@
-from pandas.util._decorators import Appender, Substitution, cache_readonly  # noqa
+import importlib as _importlib
```
Comment (jreback) [inline pandas/util/__init__.py]: umm why don't you actually just leave pandas.util.testing as a module itself? (and just re-import what you need). This is how we did all of the original deprecations.

**Review by jreback — CHANGES_REQUESTED** [review]

**Comment by jorisvandenbossche** [comment]
I don't have a strong opinion on this approach vs just leaving the pandas.util.testing module and importing what is needed there (and add a deprecation warning in the module, which will only trigger a warning once?). But if we keep the approach in this PR, should this be factored out in a utility and just replace the current DeprecatedModule utility?

**Comment by TomAugspurger** [comment]
Sorry, I think @jreback is right. We don't need the import hook stuff. Just

```python
# pandas/util/testing.py
warnings.warn("pandas.util.testing is deprecated. Use pandas.testing instead")
from pandas._testing import *
```

should accomplish the same thing with less hassle. Will update shortly (I do think I need to restructure a few things to avoid a circular import from core)

**Comment by TomAugspurger** [comment]
Sorry about that. Just using standard stuff now.

**Review by jorisvandenbossche — APPROVED** [review]

**Comment by jbrockmendel** [comment]
LGTM

File: pandas/tests/api/test_api.py [hunk 2]
```diff
@@ -288,14 +289,20 @@ def test_testing(self):
         self.check(testing, self.funcs)
 
     def test_util_testing_deprecated(self):
-        s = pd.Series([], dtype="object")
-        with tm.assert_produces_warning(FutureWarning) as m:
-            import pandas.util.testing as tm2
+        # avoid cache state affecting the test
+        sys.modules.pop("pandas.util.testing", None)
```
Comment (jreback) [inline pandas/tests/api/test_api.py]: in theory this should be in a context manager (but nbd)

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @TomAugspurger

**Pull request merged**